client/src/components/game/gui/panels/chat/ChatPanel.jsx

```diff
@@ -0,0 +1,69 @@
+import React, { useState, useRef, useEffect } from "react";
+import PropTypes from "prop-types";
+import "./ChatPanel.scss";
+import PubSub from "pubsub-js";
+import { ApplicationState, GUIState, PlayerState } from "../../../../../shared/state/States";
+import { PANEL_CHANGE } from "../../../../../shared/EventTypes";
+import Panels from "../PanelsEnum";
+
+function ChatPanel({ onCloseCallback }) {
+    const [chats, sendChat] = useState([]);
+    const chatContentsRef = useRef(null);
+    const chatInputRef = useRef(null);
+
+    useEffect(() => {
+        const subs = [
+            PubSub.subscribe(PANEL_CHANGE, () => {
+                console.log("panel change from chats");
+                if (GUIState.activePanel === Panels.Chat) {
+                    chatInputRef.current.focus();
+                }
+            }),
+        ];
+
+        return () => {
+            subs.forEach((sub) => {
+                PubSub.unsubscribe(sub);
+            });
+        };
+    });
+
+    const scrollChatToBottom = () => {
+        chatContentsRef.current.scrollTop = chatContentsRef.current.scrollHeight;
+    };
+
+    const sendC = (e) => {
+        if (e.key === "Enter") {
+            const text = e.target.value;
+            if (!text) return;
+
+            const newChat = `${PlayerState.displayName}: ${text}`;
+            sendChat((prevChats) => [...prevChats, newChat]);
+
+            ApplicationState.connection.sendEvent("chat", text);
+
+            e.target.value = "";
+
+            // add some delay to properly scroll down to edge of chats
+            setTimeout(scrollChatToBottom, 10);
+        }
+    };
+
+    return (
+        <div className="chat-container gui-zoomable">
+            <div className="chat-contents" ref={chatContentsRef}>
+                { chats.map((chat) => (<p className="chat-text" key={chat + Math.random() * 100000}>{chat}</p>)) }
```

Comment:
Using the random number to satisfy the React unique key is a bad idea, as it rerenders every chat line every time the `chats` state is set, due to the key being computed on demand, so it will be different every time and React will see it as a different child element (I figured this out the hard way... >_> ).

The better thing to do would be create something like a `ChatMessage` class you can instantiate every time a message is received, and add the random number key in there as a property, add the new instance to the `chats` list, and use it as the React key `key={chat.key}` so it won't be creating a new key each time.

If that makes sense.

Reply:
Yes yes, this is more of a quick dummy implementation, I've never used `Math.random()` as key in my react component. Will change this in later commits.